docs/source/notes/distributed_autograd.rst

```diff
@@ -0,0 +1,278 @@
+
+The first step the autograd engine performs as part of the backward pass is 
+computing the number of dependencies for each node in the autograd graph. This 
+helps the autograd engine know when a node in the graph is ready for execution.
+The numbers in brackets for ``add(1)`` and ``mul(0)`` denote the number of 
+dependencies. As you can see, this means during the backward pass the ``add`` 
+node needs 1 input and the ``mul`` node doesn't need any inputs (in other 
+words doesn't need to be executed). The local autograd engine computes these 
+dependencies by traversing the graph from the root nodes (``d`` in this case).
+
+The fact that certain nodes in the autograd graph might not be executed in the 
+backward pass poses a challenge for distributed autograd. Consider this piece 
+of code which uses RPC.
+
+.. code::
+
+  import torch
+  import torch.distributed.rpc as rpc
+
+  a = torch.rand((3, 3), requires_grad=True)
+  b = torch.rand((3, 3), requires_grad=True)
+  c = torch.rand((3, 3), requires_grad=True)
+
+  d = rpc.rpc_sync("worker1", torch.add, args=(a, b))
+  e = rpc.rpc_sync("worker1", torch.mul, args=(b, c))
+  loss = d.sum()
+
+The associated autograd graph for the code above would be:
+
+.. image:: images/distributed_dependencies.png
+
+Computing dependencies of this distributed autograd graph is much more 
+challenging and requires some overhead (either in terms of computation or 
+network communication). 
+
+For performance sensitive applications we can avoid a 
+lot of overhead by assuming every ``send`` and ``recv`` function are valid as 
+part of the backward pass (most applications don't perform RPCs that aren't 
+used). This simplifies the distributed autograd algorithm and is much more 
+efficient, but at the cost that the application needs to be aware of the 
+limitations. This algorithm is called the `FAST mode algorithm`_ and is 
+described in detail below.
+
+In the general case it might not be necessary that every ``send`` and ``recv`` 
+function is valid as part of the backward pass. To address this, we also have 
+a `SMART mode algorithm`_ which is described in a later section.
+
+FAST mode algorithm
+-------------------
+
+The key assumption of this algorithm is that each ``send`` function has a 
+dependency of 1 when we run a backward pass. In other words, we assume we'll 
+receive a gradient over RPC from another node.
+
+The algorithm is as follows:
+
+1. We start from the worker which has the roots for the backward pass 
+   (all roots must be local).
+2. Lookup all the ``send`` functions for the current 
+   `Distributed Autograd Context`_.
+3. Compute dependencies locally starting from the provided roots and all the 
+   ``send`` functions we retrieved.
+4. After computing dependencies, kick off the local autograd engine with the 
+   provided roots.
+5. When the autograd engine executes the ``recv`` function, the ``recv`` 
+   function sends the input gradients over the wire to the appropriate worker. 
```

Comment:
Suggestion to remove the idiomatic phrase "over the wire" that's similar to comment above.